keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,615 @@
+- Skew (actualJobCreationTime-expectedJobCreationTime) - [Histogram](https://prometheus.io/docs/concepts/metric_types/#histogram)
+- Job failures - [Gauge](https://prometheus.io/docs/concepts/metric_types/#counter)
+- Job successes - [Gauge](https://prometheus.io/docs/concepts/metric_types/#counter)
+- API server errors - [Gauge](https://prometheus.io/docs/concepts/metric_types/#counter)
+- Job scheduling latency (podCreationTime - jobCreationTime) - [Histogram](https://prometheus.io/docs/concepts/metric_types/#histogram)
+- Queue depth (pending CronJob and Job entries in the Queue) - [Gauge](https://prometheus.io/docs/concepts/metric_types/#gauge)
+
+Queue depth can be surfaced from existing controller framework.
+
+### Add .status.lastSuccessfulTime
+[#issue/75674](https://github.com/kubernetes/kubernetes/issues/75674)  
+Add `lastSuccessfulTime` to `.status` that tracks the last time the job completed successfully. This will augment the `lastScheduledTime` available in the `.status` in the v1beta1 api. Potential use is in monitoring (e.g. fire an alert if lastSuccessfulTime is more than X ago).
+
+
+### Add .status.nextScheduleTime
+[#issue/78564](https://github.com/kubernetes/kubernetes/issues/78564)
+Add `nextScheduleTime` to `.status` that tracks the next time the job will be scheduled. This may not be accurate with `Forbid` concurrency policy. This only tracks the `Job` creation time and not the actual `Pod` creation time.
+
+| Concurrency policy | notes |
+|:----------|:----------|
+| Allow   | `nextScheduleTime` would be accurate within a margin of controller scheduling jitter |
+| Forbid  |  `nextScheduleTime` would not be accurate if the previous `Job` takes longer than the cron interval time |
+| Replace | `nextScheduleTime` would be accurate within a margin of controller scheduling jitter along with older concurrent `Job` cleanup if applicable. |
+
+
+
+### Add Counters
+These counters would be added to `.status` section of the CronJob object:
+- `SuccessfulRuns` Count of all successful runs
+- `FailedRuns` Count of all failed runs
+- `FailuresSinceSuccess` Count of failed runs since last successful
+
+
+
+### Add .status.conditions
+Add a condition array with `Settled` condition type. This would help with the effort of standarizng conditions across all core types. `Settled` is set at the end of every successful reconcile run. The key thing to note here is the notions of `Settled` does not imply the `Job`s are running correctly. It just means that the controller is done processing this object successfully.
+
+
+### Support Jitter for cronjobs
+We propose to introduce `.spec.jitter` which is a percentage of the time delta to the next schedule. We propose to cap it to 50%. There is also a [community request](https://github.com/kubernetes/community/issues/2440) for this.
+
+```golang
+delta = nextScheduleTime - currentTime
+jitter = delta*cronjob.Spec.Jitter/100
+nextScheduleTime += jitter
+```
+
+### Support Timezone for cronjobs
```

Comment:
Suggestion: this may be better suited for a v2. This feels easy to get wrong, as a number of other commenters have suggested. It would be unfortunate to rush a design to v1 while trying to promote the other improvements.

Reply:
> It would be unfortunate to rush a design to v1 while trying to promote the other improvements.

If rushing is a concern, I'd prefer to extend the beta. Add `.spec.timezone` along with other tidying, and try that out for a release cycle (or more).

Reply:
I do see a use case for this. Especially with multi cluster scenarios. Having a schedule with local timezone or UTC (depending on master VM/pod) would be good for most use cases. Having a deterministic schedule may be preferable for others. 

But i do agree we need to get some consensus on the priority and when it needs to be done.

Reply:
I would suggest moving it to a separate KEP. I agree it's potentially controversial, while the overall idea behind this KEP isn't. I don't want to have this KEP stuck for too long (it's already hanging too long).